feat: Implement debounce function for VideoSearch component

SangJin983 opened the pull request:
### 변경 사항 (리뷰 반영)
1. videoSlice에서 불필요한 전역 상태(searchQuery) 제거
2. fetchVideos 네이밍 수정 (fetchVideos -> setVideos)
3. VideoSearch에서 적절한 상태 관리를 통한 데이터 흐름 처리 개선
4. VideoList의 불필요한 상태와 useEffect 제거
5. 불필요한 debounce의 적용 제거
6. debounce의 타이머 관리 변경
7. VideoSearch에서의 불분명한 상태명 수정 (query -> searchTerm)
8. getRandomQuery 네이밍 수정 (getRandomQuery -> generateRandomSearchTerm)

### 피드백 설명

1. 기존의 videoSlice의 불필요한 전역 상태(searchQuery)를 제거했습니다. 
2. 기존의 videoSlice 내의 fetchVideos의 잘못된 네이밍을 setVideos로 수정했습니다.
3. VideoSearch에서 사용자의 입력에 따라 데이터를 직접 가져오고 리덕스에 상태를 업데이트하는 식으로 데이터 흐름을 올바르게 수정했습니다.
4. VideoList에서 불필요한 전역 상태와 useEffect 사용으로 작성되었던 코드를 제거하였습니다. useSelector를 통한 랜더링 작업만 집중 할 수 있게 컴포넌트를 개선했습니다.
5. debounce의 경우, 현재 videoSearch에서는 handleSubmit이 호출될 경우, searchTerm을 초기화하기 때문에 (searchTerm의 유무로 handleSubmit을 호출하기 때문에) debounce가 불필요하다고 생각하여, debounce의 적용을 취소하였습니다.
6. debounce에서 timer를 불필요하게 WeakMap으로 관리하던 것을 let timer로 수정하였습니다.
7. VideoSearch에서의 불분명한 지역 상태명인 query를 searchTerm으로 수정하여, 가독성을 개선하였습니다.
8. getRandomQuery의 네이밍을 generateRandomSearchTerm으로 수정하여 함수의 목적을 좀더 명확히 하도록 개선했습니다.

### 앞으로의 개선
- 추후에는 적절한 initialCommit 커밋을 통해 적합한 단위의 PR을 구성하고, 각 작업 단위를 보다 의미있게 만들겠습니다.

---

### 변경 사항 (최초 PR)
- debounce 함수 구현
- VideoSearch 컴포넌트에 debounce 함수 적용

### 문제 설명
README에서 요구하는 debounce 함수를 구현했습니다. 사용자가 검색어 입력 후, 연속으로 submit을 시도할 경우 발생할 수 있는 불필요한 호출을 방지했습니다.

### 기타 정보
- 관련 문서: [Lodash의 debounce 설명](https://lodash.com/docs/4.17.15#debounce)

SangJin983 committed:
feat: debounce 기능 추가 및 VideoSearch에 적용

SangJin983 committed:
refactor: debounce함수의 유틸리티 재사용성과 timer관리 개선
- VideoSearch 컴포넌트에서 utils 파일로 debounce 함수를 옮겼다.
- let timer로 관리하던 방식을 WeakMap으로 바꿨다.

pjb6510 commented on src/utils/utils.js:
```diff
@@ -4,3 +4,18 @@ export const getRandomQuery = () => {
   return queries[randomIndex];
 };
+
+export const debounce = (fn, ms) => {
+  const timers = new WeakMap();
+
+  return (...args) => {
+    if (timers.has(fn)) {
+      clearTimeout(timers.get(fn));
+    }
+    const timer = setTimeout(() => {
+      fn(...args);
+      timers.delete(fn);
+    }, ms);
+    timers.set(fn, timer);
+  };
+};
```
WeakMap은 불필요합니다.
그냥 timer를 let으로 설정하고 clearTimeout 하면 됩니다.

만약 WeakMap이 debounce 함수 스코프 밖에 있다고 한다면 debounce할 여러 함수를 각각 관리하기 위해 WeakMap이 필요할 수 있지만, 지금과 같은 경우 debounce는 호출할 때마다 새로운 WeakMap 인스턴스를 생성할 것이고 여기서 WeakMap은 필연적으로 debounce될 함수 하나만 관리하게 됩니다.

pjb6510 commented on src/components/VideoSearch.jsx:
```diff
@@ -1,17 +1,25 @@
 
 const VideoSearch = () => {
   const dispatch = useDispatch();
   const [query, setQuery] = useState("");
```
query 변수명 이상합니다. query DB에 날리는 질의문을 의미하는데 여긴 그냥 검색어로 보여집니다.

pjb6510 commented on src/components/VideoSearch.jsx:
```diff
@@ -1,17 +1,25 @@
 
+  const debouncedSetSearchQuery = debounce((searchQuery) => {
+    dispatch(setSearchQuery(searchQuery));
+  }, 500);
+
+  const debouncedResetQuery = debounce(() => {
+    setQuery("");
+  }, 500);
+
   const handleSubmit = (e) => {
     e.preventDefault(); // 새로고침 방지
     if (query.trim()) {
-      dispatch(setSearchQuery(query));
-      setQuery(""); // 입력필드 초기화
+      debouncedSetSearchQuery(query);
+      debouncedResetQuery();
     }
   };
```
전체 코드를 보지 않아서 searchQuery랑 그냥 query랑 각각 무슨 역할을 하는지는 모르겠는데 debounce를 굳이 따로따로 해야하나 싶습니다.

pjb6510 replied:
방금 코드 보고 왔는데, redux로 관리되는 searchQuery는 불필요한 전역상태로 보입니다.

현재 플로우를 보면,
submit -> setSearchQuery -> useEffect(fetchVideo)[searchQuery] 방식으로 되어있는데 이는 전역상태 값에 대한 명백한 오용입니다.
react에서 '상태'는 1차적으로 sideEffect가 아니라 렌더링을 주목적으로 사용된다는 점을 명심하십시오.
따라서 redux는 fetchVideos로 가져온 videos를 들고 있는 것이 적절합니다.

이는 아래와 같은 플로우가 되어야 합니다.
submit -> handleSubmit(fetchVideos) -> [redux]: setVideos(videos)
  -> VideoList에서 redux videos를 사용해서 VideoItem 렌더링

pjb6510 reviewed (changes requested):
수정해보십시오.

그리고 앞으로는 적절히 정리된 템플릿을 initialCommit 으로 둔 후
커밋의 작업단위를 고려하여, 의미있게 작업한 코드가 PR diff로 나타날 수 있도록 PR을 만드세요.
이처럼 중간 작업이 뜬금없이 나타나면 안됩니다.

작업단위를 적절하게 잘라서 PR을 예쁘게 만드는 것은 협업의 기본입니다.

SangJin983 committed:
refactor: 리뷰 피드백을 반영한 올바른 리덕스 상태 관리를 적용 후 각 컴포넌트에 반영
- videoSlice에서 불필요한 전역상태 제거했습니다.
- fetchVideos를 적절한 네이밍인 setVideos로 수정했습니다.
- VideoSearch에서 사용자가 입력한 검색어를 바탕으로 직접 fetch후 setVideos를 호출하는 방식으로 데이터 흐름을 수정했습니다.
- VideoList가 items만을 가지고 필요한 랜더링 작업에 집중하게 수정했습니다.
- 불필요한 debounce의 타이머 관리와 적용을 수정했습니다.
- VideoSearch에서의 불분명한 상태명인 query를 searchTerm으로 수정했습니다.
- getRandomQuery의 네이밍을, generateRandomSearchTerm으로 수정했습니다.

pjb6510 reviewed (approved):
👍
